extras/axelar-token/src/IpcTokenHandler.sol

```diff
@@ -52,13 +53,38 @@ contract IpcTokenHandler is InterchainTokenExecutable, IpcHandler {
         address tokenAddr,
         uint256 amount
     ) internal override {
-        (SubnetID memory subnet, address recipient) = abi.decode(data, (SubnetID, address));
-
         IERC20 token = IERC20(tokenAddr);
         require(token.balanceOf(address(this)) >= amount, "insufficient balance");
 
         // Authorize the IPC gateway to spend these tokens on our behalf.
-        token.approve(address(_ipcGateway), amount);
+        token.safeIncreaseAllowance(address(_ipcGateway), amount);
+
+        // To be extra safe, we wrap the decoding of parameters and the interaction with the IPC gateway
+        // in a low-level call, so we can detect when any of those revert and send the funds to the admin account.
+        (bool success, ) = address(this).call(
+            abi.encodeWithSignature("_fallibleFundSubnet(bytes,uint256)", data, amount)
+        );
+
+        if (!success) {
+            // Restore the original allowance.
+            token.safeDecreaseAllowance(address(_ipcGateway), amount);
+
+            // Increase the allowance of the admin address so they can retrieve these otherwise lost tokens.
+            token.safeIncreaseAllowance(_admin, amount);
+
+            // Emit a FundingFailed event; we can't associate a specific subnet or recipient since parsing may have failed.
+            SubnetID memory nilSubnet;
+            emit FundingFailed(nilSubnet, address(0), amount);
```

Comment:
Can we want to emit msg.sender or tx.origin here? Or will that tend to be not helpful information like "gateway" or "ipc subnet" would be the callers?

Reply:
I refactored the solution a bit, and now we're able to pass actual data here.